### PR TITLE
fix: some improvements to GET chunk writing

### PR DESCRIPTION
**Pull request by dordsor21**
- ensure levelChunk is loaded before giving to copy GET - this is not necessarily guaranteed to be nonnull if two edits overlap. Whilst not advised, such an easy failure should not occur when two edits collide
 - aims to help #2729

### CODE REVIEW

**Review by NotMyFault — APPROVED** [review]

**Comment by dordsor21** [comment]
fwiw I threw this at someone who experienced fewer issues after installing

File: worldedit-bukkit/src/main/java/com/fastasyncworldedit/bukkit/adapter/NMSAdapter.java [hunk 1]
```diff
@@ -140,4 +144,122 @@ public void sendChunk(IChunkGet chunk, int mask, boolean lighting) {
...
+            lock.writeWaiting = true;
+            return lock;
+        });
+        try {
+            if (stamp[0] == 0) {
+                stamp[0] = chunkLock[0].lock.writeLock();
+            }
+            return ReflectionUtils.compareAndSet(sections, expected, value, layer);
+        } finally {
+            chunks = SENDING_CHUNKS.get(worldName);
```
Comment (SirYwell) [inline worldedit-bukkit/src/main/java/com/fastasyncworldedit/bukkit/adapter/NMSAdapter.java:195]: is there any scenario where this is actually needed? The old reference should still be the correct one as we never remove it. Alternatively, did you consider using a `ConcurrentHashMap<*WorldIntPair*, ChunkLock>` instead? This way we won't have unused CHM instances laying around (which could probably be a problem with many temporary worlds).

File: worldedit-bukkit/src/main/java/com/fastasyncworldedit/bukkit/adapter/NMSAdapter.java [hunk 2]
```diff
@@ -140,4 +144,122 @@ public void sendChunk(IChunkGet chunk, int mask, boolean lighting) {
...
+            LevelChunkSection[] sections,
+            LevelChunkSection expected,
+            LevelChunkSection value,
+            int layer
+    ) {
+        if (layer < 0 || layer >= sections.length) {
+            return false;
+        }
+        long[] stamp = new long[1];
+        ChunkLock[] chunkLock = new ChunkLock[1];
```
Comment (SirYwell) [inline worldedit-bukkit/src/main/java/com/fastasyncworldedit/bukkit/adapter/NMSAdapter.java:173]: Could these two arrays be replaced by a class? Iirc arrays have 16 byte object headers alone in current Hotspot versions, so reducing to a class would significantly cut down memory per call here (assuming it doesn't get scalar replaced, which is rather unlikely due to the complexity of the involved code).

File: worldedit-bukkit/src/main/java/com/fastasyncworldedit/bukkit/adapter/NMSAdapter.java [hunk 3]
```diff
@@ -140,4 +144,122 @@ public void sendChunk(IChunkGet chunk, int mask, boolean lighting) {
...
+    /**
+     * Releases the read lock acquired when sending a chunk packet for a chunk
+     *
+     * @since TODO
+     */
+    protected static void endChunkPacketSend(String worldName, IntPair pair, long stamp, StampedLock stampedLock) {
+        ConcurrentHashMap<IntPair, ChunkLock> chunks = SENDING_CHUNKS.get(worldName);
+        chunks.computeIfPresent(pair, (k, lock) -> {
+            if (lock.lock != stampedLock) {
+                throw new IllegalStateException("SENDING_CHUNKS stored lock is not ");
```
Comment (SirYwell) [inline worldedit-bukkit/src/main/java/com/fastasyncworldedit/bukkit/adapter/NMSAdapter.java:250]: seems like the message got messed up

**Review by SirYwell — CHANGES_REQUESTED** [review]

File: worldedit-bukkit/src/main/java/com/fastasyncworldedit/bukkit/adapter/NMSAdapter.java (hunk 1 above)
Comment (dordsor21) [reply worldedit-bukkit/src/main/java/com/fastasyncworldedit/bukkit/adapter/NMSAdapter.java:195]: The idea was to reduce lock contention between worlds when locking/unlocking. I suppose the operations themselves shouldn't take too long, though perhaps it's worth having the CHM in a BukkitWorld instance so it's GCed as needed if still attempting to do this

Comment (SirYwell) [reply worldedit-bukkit/src/main/java/com/fastasyncworldedit/bukkit/adapter/NMSAdapter.java:194]: Moving it into BukkitWorld sounds good.

Comment (dordsor21) [reply worldedit-bukkit/src/main/java/com/fastasyncworldedit/bukkit/adapter/NMSAdapter.java:194]: FAWE does not actually cache/limit the number of BukkitWorld instances, and given upstream does not either, it feels quite wrong to do this. I don't know if it's worth wrapping a WeakHashMap into a synchronised map and using the NMS world (or a normal bukkit world) as a key?

**Review by SirYwell — APPROVED** [review]

**Review by CyberFlameGO — APPROVED** [review]

**Pull request merged**